### PR TITLE
add watch mode

### LINKED ISSUE #477: Add watch mode
**Is your feature request related to a problem? Please describe.**
App developers spend quite some time context switching between developing and running cht-conf commands to update their test instances with their changes

**Describe the solution you'd like**
A way to automate running cht-conf commands in realtime when building apps

### PR DESCRIPTION
**Pull request by freddieptf**
# Description

Watches a projects directory for changes in files and automatically runs cht-conf commands to update the state of your local instance

- [x] Form Edits - changes to a form's XLSX or XML files will trigger an upload of the form to your running instance
- [x] Form deletes - deletions of a form's XLSX and XML will trigger a delete of the forms on the server 
- [x] Form Media Dirs -  we run upload app form when we detect changes to a form's media dir
- [x] Translations - we upload translations when we detect changes to any of the files under `translations`
- [x] Resources -  we upload resources when we detect changes to any of the files under `resources` or `resources.json`
- [x] App Config - we compile and upload settings when changes are made to files under app settings dir or `app_settings.json`
- [x] Top level JS files - we compile and upload settings when changes are made to a project's top level JS files

~~Unit tests are pending as i figure out how to test most of this changes~~

Fixes https://github.com/medic/app-services-team/issues/17
Fixes #477

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: src/fn/watch-project.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        }
+    };
+};
+
+let appFormListenerWait = false;
+const appFormListener = async (_, fileName) => {
+    if (appFormListenerWait) return;
+    appFormListenerWait = setTimeout(() => { appFormListenerWait = false; }, debounceDelay);
+
+    if (fileName.match(formXLSRegex)) {
```
Comment (derickl) [inline src/fn/watch-project.js:54]: we should also listen to file changes in the <form-name>-media directories and upload the corresponding form

**Comment by freddieptf** [comment]
Got stuck at making the request to the running [pouchdb](https://github.com/medic/cht-conf/blob/8f68c57ec031d3ed7fda7a80e56dff54f8925ec2/test/api-stub.js#L26). As i understood it, there should be a design doc on the the db with a rewrite rule for app_settings based on this url https://github.com/medic/cht-conf/blob/8f68c57ec031d3ed7fda7a80e56dff54f8925ec2/src/lib/api.js#L39 so i created a design doc in pouchdb so i could copy the rule but when i checked on a local running instance all i saw under rewrites is an entry with `manifest.appcache` which i really don't know where/what its used for. Ended up mocking the thing https://github.com/freddieptf/cht-conf/blob/a557e0198c3442734c43731bab0c2f093353eadc/test/fn/watch-project.spec.js#L29-L46. Will have to clean up the callback approach. WIll also have to figure out why the test does not run on CI

File: src/fn/watch-project.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            watchers.push(watchFormMediaDir(fileName, absDirPath, projectPath));
+        });
+        watchers.push(fs.watch(contactFormsPath, contactFormListener(projectPath, callback)));
+        [
+            projectPath,
+            path.join(projectPath, 'app_settings'),
+            path.join(projectPath, 'resources'),
+            path.join(projectPath, 'translations'),
+        ].forEach((path) => {
+            if (!checkExists(path)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:139]: I believe we support configurations that: 
1. don't use `app_settings` folder and instead keep all settings in `app_settings.json`
2. projects that don't have resources
3. projects that don't use translations

I don't think the command should error out if these folders don't exist, instead it should just skip (and start watching if they are added). When you run `cht-conf` without parameters, it doesn't fail if you don't have resources or translations.

File: src/fn/watch-project.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                error('make sure', path, 'exists. You can use initialise-project-layout for new projects to get the correct project layout');
+                process.exit(1);
+            }
+            watchers.push(fs.watch(path, changeListener(projectPath, api, callback)));
+        });
+        info('watching', projectPath, 'for changes');
```
Comment (dianabarsan) [inline src/fn/watch-project.js:145]: I kind of expected the "watch" command to do an upload and then start watching.

File: src/fn/watch-project.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        if (fileName.match(formXLSRegex)) {
+            await convertContactForm(projectPath, [fileName.split('.')[0]]);
+            await uploadForms(projectPath, 'contact', {
+                id_prefix: 'contact:',
+                forms: [fileName.split('.')[0]],
+                default_context: { person: false, place: false },
+            });
+            callback(fileName, true);
+        } else {
+            if (!fileName.match(formXMLRegex)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:108]: How come the XML here is treated differently than app forms?
Again, I changed the XML directly and nothing happened (this time I didn't even get an error).

File: src/fn/watch-project.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        fs.readdirSync(appFormsPath).filter((fileName) => fileName.match(formMediaRegex)).forEach((fileName) => {
+            const absDirPath = path.join(appFormsPath, fileName);
+            watchers.push(watchFormMediaDir(fileName, absDirPath, projectPath));
+        });
+        watchers.push(fs.watch(contactFormsPath, contactFormListener(projectPath, callback)));
+        [
+            projectPath,
+            path.join(projectPath, 'app_settings'),
+            path.join(projectPath, 'resources'),
+            path.join(projectPath, 'translations'),
```
Comment (dianabarsan) [inline src/fn/watch-project.js:137]: I had an unhandled promise rejection when I added a new translation file: 
image

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Very promising! I like this new feature a lot! 
I did a first pass and left some notes and questions inline. 
I've noticed that CI didn't run over the last couple of commits. Please let me know if this is something that needs dev intervention. 

Thanks!

File: src/fn/watch-project.js [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        if (!checkExists(appFormsPath) || !checkExists(contactFormsPath)) {
+            error('make sure', projectPath, 'has a valid project layout. You can use initialise-project-layout for new projects to get the correct project layout');
+            process.exit(1);
+        }
+        watchers.push(fs.watch(appFormsPath, appFormListener(projectPath, callback)));
+        fs.readdirSync(appFormsPath).filter((fileName) => fileName.match(formMediaRegex)).forEach((fileName) => {
+            const absDirPath = path.join(appFormsPath, fileName);
+            watchers.push(watchFormMediaDir(fileName, absDirPath, projectPath));
+        });
+        watchers.push(fs.watch(contactFormsPath, contactFormListener(projectPath, callback)));
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/fn/watch-project.js]: I'm considering https://www.npmjs.com/package/cross-platform-file-watch which normalizes the emitted events across different OSes

[the remark this answers is not in the document]
Comment (freddieptf) [reply src/fn/watch-project.js]: Settled on https://www.npmjs.com/package/@parcel/watcher

**Comment by freddieptf** [comment]
@dianabarsan as you had mentioned not sure why CI stopped running https://github.com/medic/cht-conf/actions?query=branch%3Awatch_mode

**Comment by derickl** [comment]
Could CI not be running since the PR is from outside the Medic organisation?

**Comment by dianabarsan** [comment]
@derickl I'm not sure, I would say no, because it ran for some of the previous commits: https://github.com/medic/cht-conf/runs/5192296825?check_suite_focus=true

**Comment by dianabarsan** [comment]
Could you please try to resolve the conflicts?

**Comment by freddieptf** [comment]
I've rebased off upstream and the actions are running now

**Review by dianabarsan — COMMENTED** [review]
Cool, please let me know when test have passed and this is ready for review.

**Comment by freddieptf** [comment]
@derickl yes but i tested locally it still runs on versions 8.x, the issue was actually how i instantiated the promise queue library i added. Had to move it to the function scope for it to work correctly on 8.x

File: src/fn/watch-project.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+const watcher = require('@parcel/watcher');
+const convertForms = require('../lib/convert-forms');
+const uploadForms = require('../lib/upload-forms');
+const convertContactForm = require('./convert-contact-forms').execute;
+const { uploadAppSettings } = require('./upload-app-settings');
+const compileAppSettings = require('./compile-app-settings').execute;
+const uploadCustomTranslations = require('./upload-custom-translations').execute;
+const uploadResources = require('./upload-resources').execute;
+
+const formXLSRegex = /^[a-zA-Z0-9_-]+\.xlsx$/;
```
Comment (dianabarsan) [inline src/fn/watch-project.js:15]: I think these regexes are a bit risky. For example, the code we use to fetch which files to convert in `conver-forms` only checks for the extension. I think we should find a way to reuse that code here as well, so there are no differences. 
Another idea, similar to the other comment I added, would be for the "fn" itself to export the extension to watch.

File: src/fn/watch-project.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+const formPropertiesRegex = /^[a-zA-Z0-9_-]+\.properties.json$/;
+const formMediaRegex = /^[a-zA-Z0-9_]+(?:-media)$/;
+const formXMLRegex = /^[a-zA-Z0-9_-]+\.xml$/;
+
+const uploadInitialState = (api) => {
+    return Promise.all(
+        [
+            uploadResources(),
+            uploadForms(environment.pathToProject, 'app'),
+            uploadForms(environment.pathToProject, 'contact', { id_prefix: 'contact:', default_context: { person: false, place: false } }),
```
Comment (dianabarsan) [inline src/fn/watch-project.js:25]: Can we just call the `uploadContactForms.execute` and `uploadAppForms.execute` here to avoid this duplication?

File: src/fn/watch-project.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+const watchProject = {
+    watch: async (api, blockFn, callback = {}, uploadStateOnStart = false) => {
+        if (uploadStateOnStart) {
+            await uploadInitialState(api).then(() => info('Initial State Uploaded'));
+        }
+
+        const appFormsPath = path.join(environment.pathToProject, 'forms', 'app');
+        const contactFormsPath = path.join(environment.pathToProject, 'forms', 'contact');
+        eventQueue =  new Queue({ concurrent: 1, start: true });
```
Comment (dianabarsan) [inline src/fn/watch-project.js:57]:
```suggestion
        eventQueue = new Queue({ concurrent: 1, start: true });
```

File: src/fn/watch-project.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    );
+};
+
+let fsEventsSubscription;
+let eventQueue;
+
+const cleanUp = async () => {
+    if (fsEventsSubscription) await fsEventsSubscription.unsubscribe();
+    if(eventQueue) await eventQueue.clear();
+};
```
Comment (dianabarsan) [inline src/fn/watch-project.js:38]: For safety, please also unset `fsEventsSubscription` and `eventQueue` here. 
```suggestion
    if (fsEventsSubscription) {
        await fsEventsSubscription.unsubscribe();
        fsEventsSubscription = undefined;
    }
    if(eventQueue) {
        await eventQueue.clear();
        eventQueue = undefined; 
    }
```

File: src/fn/watch-project.js [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+            await cleanUp();
+            resolve();
+        });
+    });
+};
+
+const watchProject = {
+    watch: async (api, blockFn, callback = {}, uploadStateOnStart = false) => {
+        if (uploadStateOnStart) {
+            await uploadInitialState(api).then(() => info('Initial State Uploaded'));
```
Comment (dianabarsan) [inline src/fn/watch-project.js:52]: Please avoid using `await ... then` 
```suggestion
await uploadInitialState(api);
info('Initial State Uploaded')
```

File: src/fn/watch-project.js [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+        const contactFormsPath = path.join(environment.pathToProject, 'forms', 'contact');
+        eventQueue =  new Queue({ concurrent: 1, start: true });
+
+        fsEventsSubscription = await watcher.subscribe(environment.pathToProject, async (err, events) => {
+            if (err) {
+                error(err);
+                return;
+            }
+            for (const event of events) {
+                if (!(event.type === 'update' || event.type === 'create')) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:65]: Please avoid negating whole blocks, it makes the condition harder to read. 
```suggestion
if (event.type !== 'update' && event.type !== 'create') {
```

File: src/fn/watch-project.js [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+                    continue;
+                }
+                const changePath = event.path;
+                const parsedPath = path.parse(changePath);
+
+                if (parsedPath.dir.startsWith(appFormsPath)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:71]: Please extract this whole block to a separate function.

File: src/fn/watch-project.js [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+                    const fileName = parsedPath.base;
+                    if (fileName.match(formXMLRegex) || fileName.match(formPropertiesRegex)) {
+                        eventQueue.enqueue(() => {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:74]: Can this be made async?
```suggestion
eventQueue.enqueue(async () => {
                            await uploadForms(environment.pathToProject, 'app', { forms: [fileName.split('.')[0]] });
                            return fileName;
```

File: src/fn/watch-project.js [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+                            return uploadForms(environment.pathToProject, 'app', {
+                                forms: [fileName.split('.')[0]],
+                            }).then(() => Promise.resolve(fileName));
```
Comment (dianabarsan) [inline src/fn/watch-project.js:77]: What is the reason for returning the file name here? I don't see its purpose in this file.

File: src/fn/watch-project.js [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+                        });
+                        continue;
+                    }
+                    if (fileName.match(formXLSRegex)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:81]: Please extract this in a separate function.

File: src/fn/watch-project.js [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+                        eventQueue.enqueue(() => {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:82]: Please make this function async. It's cleaner code to stick to one "promise" paradigm and avoid a mix and match of async and thens.

File: src/fn/watch-project.js [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+                            return convertForms(environment.pathToProject, 'app', {
+                                enketo: true,
+                                forms: [fileName.split('.')[0]],
+                                transformer: xml => xml.replace('</instance>', '</instance>\n      <instance id="contact-summary"/>'),
+                            }).then(() => Promise.resolve(fileName));
+                        });
+                        continue;
+                    }
+                    const formMediaDir = parsedPath.dir.replace(appFormsPath, '');
+                    if (formMediaDir && formMediaDir.match(formMediaRegex)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:92]: Please extract this in a separate function.

File: src/fn/watch-project.js [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+                            }).then(() => Promise.resolve(fileName));
+                        });
+                        continue;
+                    }
+                    if (fileName.match(formXLSRegex)) {
+                        eventQueue.enqueue(() => {
+                            return convertForms(environment.pathToProject, 'app', {
+                                enketo: true,
+                                forms: [fileName.split('.')[0]],
+                                transformer: xml => xml.replace('</instance>', '</instance>\n      <instance id="contact-summary"/>'),
```
Comment (dianabarsan) [inline src/fn/watch-project.js:86]: Can we somehow reuse the `converAppForms.execute` call here, so we avoid duplicating the transformer?

File: src/fn/watch-project.js [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+                                .then(() => Promise.resolve(fileName));
+                        });
+                        continue;
+                    }
+                    if (fileName.match(formXMLRegex)) {
+                        eventQueue.enqueue(() => {
+                            return uploadForms(environment.pathToProject, 'contact', {
+                                id_prefix: 'contact:',
+                                forms: [fileName.split('.')[0]],
+                                default_context: { person: false, place: false },
```
Comment (dianabarsan) [inline src/fn/watch-project.js:117]: As above, can we find a way to reuse the `convertContactForms.execute` here, so we don't duplicate the default_context and id_prefix params?

File: src/fn/watch-project.js [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+                    const fileName = parsedPath.base;
+                    if (fileName.match(/messages-[\w]+\.properties/)) {
+                        await uploadCustomTranslations();
+                        if (callback) callback(fileName);
+                        continue;
+                    }
+                    continue;
+                }
+
+                if ((parsedPath.dir === environment.pathToProject && parsedPath.base !== 'resources.json')
```
Comment (dianabarsan) [inline src/fn/watch-project.js:135]: this condition is very hard to read. Please store it in a variable. It would also be useful if `compileAppSettings` would export the relevant paths, or even the path matcher altogether.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Very very nice work! I left a couple of readability and maintainability related requests. 
From a lack of time, I skipped testing this and just did a read through. Also, apoligies for the delay for reviewing. :(

File: src/fn/watch-project.js (hunk 15 above)
Comment (freddieptf) [reply src/fn/watch-project.js:83]: It's what i pass to the callback when the promise is resolved. It's only used by the tests right now https://github.com/medic/cht-conf/blob/9f66570b85556aa9ef65f374b4a775da2051b818/src/fn/watch-project.js#L156-L159

File: src/fn/watch-project.js [hunk 22]
```diff
@@ -0,0 +1,177 @@
...
+    });
+};
+
+const watchProject = {
+    watch: async (api, blockFn, callback = {}, uploadStateOnStart = false) => {
+        if (uploadStateOnStart) {
+            await uploadInitialState(api).then(() => info('Initial State Uploaded'));
+        }
+
+        const appFormsPath = path.join(environment.pathToProject, 'forms', 'app');
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/fn/watch-project.js]: This is a great idea!

**Comment by freddieptf** [comment]
@dianabarsan i've had to use relative paths for the fn exported paths as the environment is not initialized. I'm not sure how clean my implementation is as in some cases we have 2 functions i.e `upload-{some resource}` and `convert-{some resource}` exporting the same path, do you think it's a good idea to create a shared file the functions can import for this scenario?

File: src/fn/watch-project.js [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+            await uploadContactForms([fileName.split('.')[0]]);
+            return fileName;
+        });
+        return true;
+    }
+    return false;
+};
+
+const processConfigFiles = (api, fileName) => {
+    if (fileName === 'app_settings.json') {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:103]: It would be good if `app_settings.json` was exported by `uploadAppSettings`.

File: src/fn/watch-project.js [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+        eventQueue.enqueue(async () => {
+            await uploadAppSettings(api);
+            return fileName;
+        });
+        return true;
+    }
+    if (fileName.match(/.+\.js$/) || fileName.match(/^[\w]+\.json$/)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:110]: It would be good if these matchers were exported by `compileAppSettings`.

File: src/fn/watch-project.js [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+                return;
+            }
+            for (const event of events) {
+                if (event.type !== 'update' && event.type !== 'create') {
+                    continue;
+                }
+                const changePath = event.path;
+                const parsedPath = path.parse(changePath);
+
+                if (parsedPath.dir.replace(environment.pathToProject, '').startsWith(`/${APP_FORMS_PATH}`)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:141]: I think this will be more readable like: 
```suggestion
if (parsed.dir === path.join(environment.pathToProject, APP_FORMS_PATH))
```

File: src/fn/watch-project.js [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+                    const relativeFormsDir = parsedPath.dir.replace(environment.pathToProject, '');
+                    const fileName = parsedPath.base;
+                    if (!processAppForm(fileName) && !relativeFormsDir.endsWith(APP_FORMS_PATH)) {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:144]: I find this is hard to follow, it's not clear that we're trying to match for form media folders. I think we should just use the regex we've already defined and have this in a separate section, just to keep code clear.

File: src/fn/watch-project.js [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+            for (const event of events) {
+                if (event.type !== 'update' && event.type !== 'create') {
+                    continue;
+                }
+                const changePath = event.path;
+                const parsedPath = path.parse(changePath);
+
+                if (parsedPath.dir.replace(environment.pathToProject, '').startsWith(`/${APP_FORMS_PATH}`)) {
+                    const relativeFormsDir = parsedPath.dir.replace(environment.pathToProject, '');
+                    const fileName = parsedPath.base;
```
Comment (dianabarsan) [inline src/fn/watch-project.js:143]: This line is repeated 6 times. Please consider moving this to the top, under the definition of parsedPath.

File: src/fn/watch-project.js [hunk 28]
```diff
@@ -0,0 +1,206 @@
...
+                if (parsedPath.dir === path.join(environment.pathToProject, CONTACT_FORMS_PATH)) {
+                    const fileName = parsedPath.base;
+                    processContactForm(fileName);
+                    continue;
+                }
+
+                if (parsedPath.dir === path.join(environment.pathToProject, TRANSLATIONS_DIR_PATH)) {
+                    await uploadCustomTranslations();
+                    const fileName = parsedPath.base;
+                    if (callback) callback(fileName);
```
Comment (dianabarsan) [inline src/fn/watch-project.js:160]: It's unclear why the callback is only called for some actions, and not called for others. 
Best practice is to design your tests so that they work with your code, and not the other way around. If this is just required for a test, can the test be refactored so it does not require this or at least be consistent with the callbacks?

File: test/fn/watch-project.spec.js [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+      .then(() => {
+        return '{"success": true, "updated": true}';
+      });
+  },
+  getAppSettings: () => api.db.get('app_settings')
+};
+
+function editBaseSettings() {
+  return new Promise((resolve) => {
+    const appSettings = fs.readJson(settingsPath);
```
Comment (dianabarsan) [inline test/fn/watch-project.spec.js:29]: The reason for these calls to be wrapped in promises is unclear on first read. 
Could you please add a comment to describe why this is needed?

File: test/fn/watch-project.spec.js [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+    expect(fs.fs.existsSync(appSettingsPath)).to.be.false;
+    return watchWrapper(editBaseSettings, 'base_settings.json')
+      .then(() => { return fs.readJson(appSettingsPath); })
+      .then((settings) => expect(settings.locale).equal('es'));
+  });
+
+  it('watch-project: upload custom translations', () => {
+    return uploadCustomTranslations()
+      .then(() => expectTranslationDocs(api, 'en'))
+      .then(watchWrapper(editTranslations, 'messages-en.properties'))
```
Comment (dianabarsan) [inline test/fn/watch-project.spec.js:123]:
```suggestion
.then(() => watchWrapper(editTranslations, 'messages-en.properties'))
```

File: test/fn/watch-project.spec.js [hunk 31]
```diff
@@ -0,0 +1,223 @@
...
+
+  it('watch-project: convert app forms', () => {
+    const appFormPath = path.join(testDir, 'forms', 'app');
+    const copyForm = () => copySampleForms('convert-app-form');
+    return watchWrapper(copyForm, 'death.xlsx')
+      .then(() => {
+        expect(fs.fs.readdirSync(appFormPath).find(file => file === 'death.xml')).to.be.not.undefined;
+      })
+      .then(() => {
+        fs.fs.readdirSync(appFormPath).filter(name => name.startsWith('death')).forEach(file => fs.fs.unlinkSync(path.join(appFormPath, file)));
```
Comment (dianabarsan) [inline test/fn/watch-project.spec.js:149]: Please consider saving `death` string as a variable and reuse it. It's used 3 times in this test.

File: test/fn/watch-project.spec.js [hunk 32]
```diff
@@ -0,0 +1,223 @@
...
+      return new Promise((resolve) => {
+        copySampleForms('upload-app-form');
+        resolve();
+      });
+    };
+    api.giveResponses({ status: 200, body: { ok: true } });
+    return watchWrapper(copySampleForm, 'death.xml')
+      .then(() => api.db.allDocs())
+      .then(docs => {
+        expect(docs.rows.filter(row => row.id === 'form:death')).to.not.be.empty;
```
Comment (dianabarsan) [inline test/fn/watch-project.spec.js:164]:
```suggestion
const docIds = doc.rows.map(row => row.id);
expect(dicIds).to.include('form:death');
```

**Review by dianabarsan — COMMENTED** [review]
Great work!! 
I left a couple of readability requests inline.

File: test/fn/watch-project.spec.js (hunk 29 above)
Comment (freddieptf) [reply test/fn/watch-project.spec.js:29]: thanks for pointing that out. These functions ares substitutes for the `watch` blocking function and i was under the impression i could only call `await` on Promises https://github.com/medic/cht-conf/blob/b715eea8a62176c82f49c25c6858b05f9ccc4716/src/fn/watch-project.js#L121

File: src/fn/watch-project.js (hunk 28 above)
Comment (freddieptf) [reply src/fn/watch-project.js:155]: It's called for all functions, the actions added to the event queue return the filename when resolved https://github.com/medic/cht-conf/blob/b715eea8a62176c82f49c25c6858b05f9ccc4716/src/fn/watch-project.js#L185-L189

Comment (freddieptf) [reply src/fn/watch-project.js:155]: I'm not sure if it's possible to run the tests without a way of getting notified of changes but let me see if i can figure out a way

File: test/fn/watch-project.spec.js [hunk 33]
```diff
@@ -0,0 +1,265 @@
...
+      .then(messages => {
+        assert.deepEqual(messages.custom, { a: 'first', test: 'new' });
+      });
+  });
+
+  it('watch-project: upload resources', () => {
+    return watchWrapper(editResources, 'resources.json')
+      .then(() => api.db.allDocs())
+      .then(docs => {
+        expect(docs.rows.filter(row => row.id === 'resources')).to.not.be.empty;
```
Comment (dianabarsan) [inline test/fn/watch-project.spec.js:131]:
```suggestion
        const docIds = docs.rows.map(row => row.id);
        expect(docIds).to.include('resources');
```

File: src/fn/watch-project.js [hunk 34]
```diff
@@ -0,0 +1,234 @@
...
+
+        eventQueue = new Queue({ concurrent: 1, start: true });
+
+        fsEventsSubscription = await watcher.subscribe(environment.pathToProject, async (err, events) => {
+            if (err) {
+                error(err);
+                return;
+            }
+            for (const event of events) {
+                if (event.type !== 'update' && event.type !== 'create') {
```
Comment (dianabarsan) [inline src/fn/watch-project.js:159]: Is watching for deletions planned for the next iteration? For example when a form is deleted.

File: src/fn/watch-project.js [hunk 35]
```diff
@@ -0,0 +1,234 @@
...
+                    continue;
+                }
+
+                if (parsedPath.dir === environment.pathToProject && parsedPath.base !== RESOURCE_CONFIG_PATH) {
+                    processConfigFiles(api, fileName);
+                    continue;
+                }
+
+                if (parsedPath.base === RESOURCE_CONFIG_PATH || parsedPath.dir === path.join(environment.pathToProject, RESOURCES_DIR_PATH)) {
+                    await uploadResources();
```
Comment (dianabarsan) [inline src/fn/watch-project.js:206]: We need to handle errors for all of these calls, right now they're unhandled rejections. 
For example, if I change the "resources.json" file and make the json invalid, I get: 
```
WARN Error parsing JSON in: /home/diana/projects/medic-utils/conf/default/resources.json 
(node:1952248) UnhandledPromiseRejectionWarning: SyntaxError: Unexpected token } in JSON at position 1502
    at JSON.parse (<anonymous>)
    at Object.readJson (/home/diana/projects/medic-conf/src/lib/sync-fs.js:29:17)
    at module.exports (/home/diana/projects/medic-conf/src/lib/upload-configuration-docs.js:47:19)
    at execute (/home/diana/projects/medic-conf/src/fn/upload-resources.js:19:12)
    at /home/diana/projects/medic-conf/src/fn/watch-project.js:206:27
(node:1952248) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:1952248) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

File: src/fn/watch-project.js (hunk 28 above)
Comment (dianabarsan) [reply src/fn/watch-project.js:191]: One option would be to watch Couchdb changes and run your assertions when detecting changes. But that't not necessarily achievable, since some actions make many changes that you'd need to keep track of...

**Review by dianabarsan — COMMENTED** [review]
Excellent progress! 
I just have one last comment about handling errors from actions when watching.

**Review by dianabarsan — COMMENTED** [review]
Fantastic work! I just have one question about consistency when deleting form "parts".

File: src/fn/watch-project.js [hunk 36]
```diff
@@ -228,7 +232,11 @@ const watchProject = {
                 }
 
                 if (parsedPath.base === RESOURCE_CONFIG_PATH || parsedPath.dir === path.join(environment.pathToProject, RESOURCES_DIR_PATH)) {
-                    await uploadResources();
+                    try {
+                        await uploadResources();
+                    } catch (e) {
+                        error(e);
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/fn/watch-project.js:238]: The other errors get caught and reported at https://github.com/medic/cht-conf/blob/f7bc7ce2431a0768714fc9263401cca63e41137d/src/fn/watch-project.js#L251
I use the queue to enforce an order to how we run some of the commands that need it but did not feel the need for order with these two specific commands. I think i'll just add them to the queue for consistency

File: src/fn/watch-project.js [hunk 37]
```diff
@@ -48,7 +55,22 @@ const waitForKillSignal = () => {
     });
 };
 
-const processAppForm = (fileName) => {
+const deleteForm = (fileName) => {
+    let form = uploadForms.formFileMatcher(fileName) || convertForms.formFileMatcher(fileName);
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/fn/watch-project.js:59]: Yes you're right, had thought of that but did not want the command running any destructive commands on the fs and kinda left it up to the user. I believe it should be safe to assume a .xls delete means we can delete the .xml on the fs too but there might be a case for an .xml delete meaning a delete on the server but we'd like to retain the .xls

File: src/fn/watch-project.js (hunk 36 above)
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/fn/watch-project.js:238]: > error('Error while uploading resources', err);

For the sake of simplicity, i'll have to make the error message generic followed by the more descriptive command error

File: src/fn/watch-project.js (hunk 37 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply src/fn/watch-project.js]: yea, I agree.
My proposal is to only delete the form on the server when both xml and xls are gone. How does that sound?

File: src/fn/watch-project.js (hunk 36 above)
[the remark this answers is not in the document]
Comment (dianabarsan) [reply src/fn/watch-project.js]: Consistency is the best approach! Please add them to the queue.

File: src/fn/watch-project.js (hunk 37 above)
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/fn/watch-project.js]: Sounds good!

**Review by dianabarsan — APPROVED** [review]
⭐ ⭐ ⭐ ⭐ ⭐

**Comment by garethbowen** [comment]
@dianabarsan Can you help @freddieptf get this ready for AT? It probably needs a cht-conf issue to track it...

**Comment by kennsippell** [comment]
Congrats @freddieptf on your first new feature in cht! May you have many more :D

**Comment by dianabarsan** [comment]
@freddieptf Could you please open an issue in the cht-conf repo that describes this feature? Next step would be to get this AT-ed and then merged.

**Comment by freddieptf** [comment]
@dianabarsan Issue created, [Issue](https://github.com/medic/cht-conf/issues/477)
